simln-lib/src/lib.rs

```diff
@@ -1548,33 +1526,184 @@ mod tests {
         );
     }
 
+    /// Tests for validate_activity()
+    #[tokio::test]
```

Comment:
These all look great!

Would like to distill the in-function comments to just one high-level comment on the function that explains what's being covered. I think the tests are readable enough (thanks to nice helper fns) to not require line-by-line comments.